lib/Collaboration/Reference/TalkReferenceProvider.php

```diff
@@ -0,0 +1,288 @@
+		 */
+		$roomName = $room->getDisplayName($this->userId);
+		$title = $roomName;
+		$description = '';
+
+		if ($participant instanceof Participant
+			|| $this->roomManager->isRoomListableByUser($room, $this->userId)) {
+			$description = $room->getDescription();
+		}
+
+
+		/**
+		 * If linking to a comment and the user is already a participant
+		 * Title is "Message of {user} in {conversation}"
+		 * Description is the plain text chat message
+		 */
+		if ($participant && !empty($referenceMatch['message'])) {
+			$comment = $this->chatManager->getComment($room, $referenceMatch['message']);
+			$message = $this->messageParser->createMessage($room, $participant, $comment, $this->l);
+			$this->messageParser->parseMessage($message);
+
+			$placeholders = $replacements = [];
+			foreach ($message->getMessageParameters() as $placeholder => $parameter) {
+				$placeholders[] = '{' . $placeholder . '}';
+				if ($parameter['type'] === 'user' || $parameter['type'] === 'guest') {
+					$replacements[] = '@' . $parameter['name'];
+				} else {
+					$replacements[] = $parameter['name'];
+				}
+			}
+			$description = str_replace($placeholders, $replacements, $message->getMessage());
+
+			$titleLine = $this->l->t('Message of {user} in {conversation}');
+			if ($room->getType() === Room::TYPE_ONE_TO_ONE) {
+				$titleLine = $this->l->t('Message of {user}');
+			}
+
+			$displayName = $message->getActorDisplayName();
+			if ($message->getActorType() === Attendee::ACTOR_GUESTS) {
+				if ($displayName === '') {
+					$displayName = $this->l->t('Guest');
+				} else {
+					$displayName = $this->l->t('%s (guest)', $displayName);
+				}
+			} elseif ($displayName === '') {
+				$titleLine = $this->l->t('Message of a deleted user in {conversation}');
+			}
+
+			$title = str_replace(
+				['{user}', '{conversation}'],
+				[$displayName, $title],
+				$titleLine
+			);
+		}
+
+		$reference->setTitle($title);
+		$reference->setDescription($description);
+		$reference->setUrl($this->urlGenerator->linkToRouteAbsolute('spreed.Page.showCall', ['token' => $room->getToken()]));
+		$reference->setImageUrl($this->getRoomIconUrl($room, $this->userId));
+
+		$reference->setRichObject('call', [
+			'id' => $room->getToken(),
+			'name' => $roomName,
+			'link' => $reference->getUrl(),
+			'call-type' => $this->getRoomType($room),
+		]);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getCachePrefix(string $referenceId): string {
+		$referenceMatch = $this->getTalkAppLinkToken($referenceId);
+		if ($referenceMatch === null) {
+			return '';
+		}
+
+		return $referenceMatch['token'];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getCacheKey(string $referenceId): ?string {
+		$referenceMatch = $this->getTalkAppLinkToken($referenceId);
+		if ($referenceMatch === null) {
+			return '';
+		}
+
+		return ($this->userId ?? '') . '#' . ($referenceMatch['message'] ?? 0);
+	}
+
+	protected function getRoomIconUrl(Room $room, string $userId): string {
```

Comment:
This method return two types of icon URL, is a good point to cover with unit test.

Reply:
The function is now only chained 3rdparty calls, so not sure tests are too useful.